cpp/include/raft/span.hpp

```diff
@@ -59,23 +59,24 @@ class span {
   /**
    * @brief Constructs a span that is a view over the range [first, first + count);
    */
-  constexpr span(pointer ptr, size_type count) noexcept : size_(count), data_(ptr)
+  constexpr span(pointer ptr, size_type count) noexcept : storage_{ptr, count}
   {
     assert(!(Extent != dynamic_extent && count != Extent));
     assert(ptr || count == 0);
   }
   /**
    * @brief Constructs a span that is a view over the range [first, last)
    */
-  constexpr span(pointer first, pointer last) noexcept : size_(last - first), data_(first)
+  constexpr span(pointer first, pointer last) noexcept
+    : storage_{first, static_cast<size_type>(thrust::distance(first, last))}
```

Comment:
This is largely just a matter of personal preference, but I would delegate ctors to the `span(pointer,size_type)` ctor. 
```suggestion
  constexpr span(pointer first, pointer last) noexcept
    : span{first, static_cast<size_type>(thrust::distance(first, last))}
```